lib/TPP/Transforms/VectorContractToOuterproduct.cpp

```diff
@@ -0,0 +1,247 @@
+//===--------------- VectorContractToOuterproduct.cpp ------------*- C++-*-===//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements lowering of vector contraction to vector outerproduct.
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Transforms/Transforms.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/Dialect/Vector/Transforms/LoweringPatterns.h"
+#include "mlir/Dialect/Vector/Transforms/VectorRewritePatterns.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Interfaces/VectorInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include <cmath>
+#include <cstdint>
+
+#define DEBUG_TYPE "vector-contract-to-outerproduct"
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_VECTORCONTRACTTOOUTERPRODUCT
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+namespace mlir {
+namespace tpp {
+
+/// Returns true if the \p map is transposed for the given operand in context of
+/// matrix multiply. \p 'isRHS = false' flag indicates thet the check is being
+/// performed on LHS operand, otherwise RHS operand.
+bool isTransposed(AffineMap map, unsigned matrixDims, bool isRHS = false) {
+  auto results = map.getResults();
+  if (results.size() != matrixDims)
+    return false;
+
+  // Check the last two dimensions for transposition
+  auto secondLast = dyn_cast<AffineDimExpr>(results[matrixDims - 2]);
+  auto last = dyn_cast<AffineDimExpr>(results[matrixDims - 1]);
+
+  if (!secondLast || !last)
+    return false;
+
+  // For LHS, If the last dimension comes before the second last, it's
+  // transposed. its opposite for RHS.
+  return isRHS ? last.getPosition() > secondLast.getPosition()
```

Comment:
You're assuming this is a 3D operation on 2D shapes of a matmul kind, but not checking any of it. Calling the function `isTransposed` doesn't help to identify what this is, and may trick future developers into using it without checking.

Here you want to guarantee the following properties:
* This function only handles matmul maps `(m, n, k)`. So, the map's input _must_ be 3D.
* It only handles 2D shapes, so the result _must_ be 2D.
* There's only 3 combinations (`m, n`, 'm, k', `k, n'). You can restrict just to inputs.

Clearly, with just inputs, there are only four options:
* `m, k` and `k, n` are non-transposed.
* `k, m` and `n, k` are transposed.

Everything else is broken. IF you just check for those, and call is `isMatmulInputTransposed`, the function will be simple to read, won't be misused and we can extend in the future if we need anything else.

Also, please make it `static`.

Reply:
One more thing. Returning `false` here has the meaning "it's not transposed", but also "there's been a problem". We need to separate them into transposed, non-transposed, failure.

For the sake of brevity, and because this will be a local static function, you can use asserts for the errors and return `true` for transposed and `false` for non-transposed. But you _have_ to make sure there will be no other returns.

Reply:
This pass is expected to work on a tiled batch_reduce_matmul too. For example:
```
#map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>
#map1 = affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>
#map2 = affine_map<(d0, d1, d2, d3) -> (d1, d2)>
module {
  func.func @batch_reduce_matmul_tiled(%arg0: memref<16x32x16x32xf32>, %arg1: memref<32x32x32x32xf32>, %arg2: memref<16x32x16x32xf32>) {
    %cst = arith.constant 0.000000e+00 : f32
    %c8 = arith.constant 8 : index
    %c1 = arith.constant 1 : index
    %c32 = arith.constant 32 : index
    %c4 = arith.constant 4 : index
    %c16 = arith.constant 16 : index
    %c0 = arith.constant 0 : index
    scf.forall (%arg3, %arg4) in (16, 32) {
      %subview = memref.subview %arg0[%arg3, 0, 0, 0] [1, 32, 16, 32] [1, 1, 1, 1] : memref<16x32x16x32xf32> to memref<32x16x32xf32, strided<[512, 32, 1], offset: ?>>
      %subview_0 = memref.subview %arg1[%arg4, 0, 0, 0] [1, 32, 32, 32] [1, 1, 1, 1] : memref<32x32x32x32xf32> to memref<32x32x32xf32, strided<[1024, 32, 1], offset: ?>>
      %subview_1 = memref.subview %arg2[%arg3, %arg4, 0, 0] [1, 1, 16, 32] [1, 1, 1, 1] : memref<16x32x16x32xf32> to memref<16x32xf32, strided<[32, 1], offset: ?>>
      scf.for %arg5 = %c0 to %c16 step %c4 {  // m-loop
        scf.for %arg6 = %c0 to %c32 step %c16 { //n-loop
          scf.for %arg7 = %c0 to %c32 step %c1 { //BRGEMM-loop
            scf.for %arg8 = %c0 to %c32 step %c8 { //k-loop
              %subview_2 = memref.subview %subview[%arg7, %arg5, %arg8] [1, 4, 4] [1, 1, 1] : memref<32x16x32xf32, strided<[512, 32, 1], offset: ?>> to memref<1x4x4xf32, strided<[512, 32, 1], offset: ?>>
              %subview_3 = memref.subview %subview_0[%arg7, %arg8, %arg6] [1, 4, 2] [1, 1, 1] : memref<32x32x32xf32, strided<[1024, 32, 1], offset: ?>> to memref<1x4x2xf32, strided<[1024, 32, 1], offset: ?>>
              %subview_4 = memref.subview %subview_1[%arg5, %arg6] [4, 2] [1, 1] : memref<16x32xf32, strided<[32, 1], offset: ?>> to memref<4x2xf32, strided<[32, 1], offset: ?>>
              %0 = vector.transfer_read %subview_2[%c0, %c0, %c0], %cst {in_bounds = [true, true, true]} : memref<1x4x4xf32, strided<[512, 32, 1], offset: ?>>, vector<1x4x4xf32>
              %1 = vector.transfer_read %subview_3[%c0, %c0, %c0], %cst {in_bounds = [true, true, true]} : memref<1x4x2xf32, strided<[1024, 32, 1], offset: ?>>, vector<1x4x2xf32>
              %2 = vector.transfer_read %subview_4[%c0, %c0], %cst {in_bounds = [true, true]} : memref<4x2xf32, strided<[32, 1], offset: ?>>, vector<4x2xf32>
              %3 = vector.contract {indexing_maps = [#map, #map1, #map2], iterator_types = ["reduction", "parallel", "parallel", "reduction"], kind = #vector.kind<add>} %0, %1, %2 : vector<1x4x4xf32>, vector<1x4x2xf32> into vector<4x2xf32>
              vector.transfer_write %3, %subview_4[%c0, %c0] {in_bounds = [true, true]} : vector<4x2xf32>, memref<4x2xf32, strided<[32, 1], offset: ?>>
            }
          }
        }
      }
    }
    return
  }
}
```
Here the map `affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>` results are 3, `d0` is the batch dimension. The last two dimensions, d1 and d3, are the one related to matmul which I to wanted to check for transposition.

Reply:
I see, so I propose we decompose the problem. 

First we detect the `batch` part, which shows 4 dimensions instead of 3. Then we detect the `reduce` part, which shows that the output has one less dimension than the inputs. Then we match against a `matmul` on the three inner-most dims in the way I expose above.

For the `matmul` part, we only need to pass the three inner-most dims, so in the `batch` case we remove the outer-dim and continue. This way the checks can be reused and we identify the shapes and dims, then apply the right outer product (or match the right XSMM uKernel).

Reply:
For now, batch matmul is not supported. Updated it as exposed.